man/vp.1
.TH vp "1" "December 15, 2023" "" "vp manual"

.SH NAME
vp - helps with development, installation and maintenance of VPM, NPM and Cargo packages

.SH SYNOPSIS
.B vp
[-c|--config
.I name
] [-l|--log
.I
file
] [-h|--heading
.I
level
]
.br
[-t|--tag-prefix
.I
prefix
] [-l|--logged-types
.I
types
] [-f|--from
.I
hash
]
.br
[-t|--to
.I
hash
] [-u|--try-unshallow] [-p|--path
.I
path
]
.br
[-r|--repo-url
.I
url
] [-o|--override-version
.I
version
]
.br
[-e|--write-changes
.I
file
] [-w|--write-version
.I
file
]
.br
[-a|--assume-patch] [-0|--bump-major-0] [--pre-release]
.br
[--pre-id
.I
id
] [-d|--dry-run]
.br
[-N|--no-failure] [-i|--print-last] [-q|--quiet] [-v|--verbose]
.br
[-V|--version] [-h|--help]
.I command
[
.I parameters
]
.RE

.SH DESCRIPTION
vp helps with development, installation and maintenance of VPM, NPM and Cargo packages.

.SH COMMANDS
.B
.IP "init"
generate a config file with defaults
.B
.IP "ls|list"
list modules installed in ~/.vmodules or ./modules
.B
.IP "link"
link the current module to the global ~./vmodules directory
.B
.IP "unlink"
remove the current module link from the global ~./vmodules
.B
.IP "version"
prepare the current module for publishing a new version
(update changelog, bump version, commit and tag the change)
.B
.IP "publish"
publish a new version prepared earlier by `vp version`
(push the change and tag, create a github release)
.B
.IP "release"
perform both `vp version` and `vp publish`

.SH OPTIONS AND PARAMETERS FOR LIST
.B
.IP "-g|--global"
inspect contents of ~./vmodules (default is ./[src/]modules)
.B
.IP "[pkg names]"
names of the package to print its version

.SH PARAMETERS FOR LINK AND UNLINK
.B
.IP "[pkg names]"
name of the package if guessing is not reliable

.SH PARAMETERS FOR VERSION AND PUBLISH
.B
.IP "[version]"
version if the changelog update is disabled
(also major, minor or patch for bumping the existing version)

.SH OPTIONS FOR VERSION, PUBLISH AND RELEASE
.B
.IP "--no-changes"
do not update the changelog
.B
.IP "--no-bump"
do not bump the version in the package manifest
.B
.IP "--no-vlang"
do not version and publish using v and v.mod
.B
.IP "--no-node"
do not version and publish using npm and package.json
.B
.IP "--no-rust"
do not version and publish using rust and Cargo.toml
.B
.IP "--no-golang"
do not version and publish using go and go.mod
.B
.IP "--no-commit"
do not commit the changes during publishing
.B
.IP "--no-tag"
do not tag the commit during publishing
.B
.IP "--no-push"
do not push the commit and tag during publishing
.B
.IP "--no-release"
do not create a new github release
.B
.IP "--no-archives"
do not upload platform archives automatically as assets
.B
.IP "--no-failure"
do not fail in case of no version change or release
.B
.IP "--no-verify"
append --no-verify to commit and push
.B
.IP "--no-npm-auth"
do not require npm token and write it to .npmrc
.B
.IP "--nc-args args"
extra arguments for newchanges, enclosed in quotes
.B
.IP "-0|--bump-major-0"
bump the major version also if it is 0
.B
.IP "pre-release"
bump the existing pre-release number
.B
.IP "pre-id id"
first pre-release identifier (default: "next")
.B
.IP "-a|--assets file"
files to upload as extra assets to the github release
.B
.IP "-b|--bump-files file"
extra files in which to bump the version
.B
.IP "-y|--yes"
answer the push and release confirmations with "yes"
.B
.IP "-d|--dry-run"
only print what would be done without doing it
.B
.IP "-v|--verbose"
print the new changes on the console too

.SH COMMON OPTIONS
.B
.IP "-c|--config name"
file name or path of the config file
.B
.IP "-V|--version"
print the version of the executable and exit
.B
.IP "-h|--help"
print the usage information and exit

.RE

.SH EXAMPLES
.B
.IP "vp link prantlf.cargs -f"
Link the current directory to ~/.vmodules/prantlf/cargs and ignore if the repository is not prantlf/cargs.
.B
.IP "vp unlink"
Remove the link to the current directory from ~/.vmodules.
.B
.IP "vp publish -v"
Publish the changes committed to the local repository with verbose output oon the console.

.SH BUGS
vp's bug tracker is located at <https://github.com/prantlf/vp/issues>

.SH DOCUMENTATION
See the website for documentation <https://github.com/prantlf/vp#readme>

.SH AUTHOR
Currently maintained by Ferdinand Prantl <prantlf@gmail.com>


CHANGELOG.md
# Changes

## [0.18.0](https://github.com/prantlf/vp/compare/v0.17.4...v0.18.0) (2025-10-11)

### Features

* Add option --no-npm-auth ([de1833f](https://github.com/prantlf/vp/commit/de1833f8a9522be51856efbefa8641eb699ead42))

## [0.17.4](https://github.com/prantlf/vp/compare/v0.17.3...v0.17.4) (2025-08-31)

### Bug Fixes

* Workaround for 422 asset uploading error ([c7d5bb8](https://github.com/prantlf/vp/commit/c7d5bb8584d294543cae0e35350d688a523f0681))

## [0.17.3](https://github.com/prantlf/vp/compare/v0.17.2...v0.17.3) (2025-08-31)

### Bug Fixes

* Publish temporarily from mac ([d6711a1](https://github.com/prantlf/vp/commit/d6711a142643b029aa85bd14c41fbeb911233e98))

## [0.17.2](https://github.com/prantlf/vp/compare/v0.17.1...v0.17.2) (2025-03-23)

### Bug Fixes

* Do not optimize binaries built on Linux ([0638b52](https://github.com/prantlf/vp/commit/0638b52a65fd800faf0dfa218dd084b3198eb66a))

## [0.17.1](https://github.com/prantlf/vp/compare/v0.17.0...v0.17.1) (2025-03-23)

### Bug Fixes

* Remove unused build configurations ([7019577](https://github.com/prantlf/vp/commit/701957705d8d19bdd024e107d32d6ed0524c4af0))

## [0.17.0](https://github.com/prantlf/vp/compare/v0.16.0...v0.17.0) (2025-03-23)

### Features

* Release Linux RISC-V archive if found ([f9a5786](https://github.com/prantlf/vp/commit/f9a57868cd59bf699b3ff61dac7e85590152f44d))

## [0.16.0](https://github.com/prantlf/vp/compare/v0.15.1...v0.16.0) (2025-03-22)

### Features

* Add RISC-V build ([c9b92f3](https://github.com/prantlf/vp/commit/c9b92f3cdba616f59389d1bd8c8079282ea23402))

### Bug Fixes

* Adapt to breaking changes in the V compiler ([5b5ba6d](https://github.com/prantlf/vp/commit/5b5ba6dbc5d260557647b6700fb850044693a29a))
* Cast integer as the new V compiler requires ([f16f921](https://github.com/prantlf/vp/commit/f16f92102cb46964b9cd879e76b57769049d8be2))

## [0.15.1](https://github.com/prantlf/vp/compare/v0.15.0...v0.15.1) (2024-11-16)

### Bug Fixes

* Fix sources for the new V compiler ([1e3738e](https://github.com/prantlf/vp/commit/1e3738eb4cba4bbd6ce00c621cb1c0797aa63eeb))

## [0.15.0](https://github.com/prantlf/vp/compare/v0.14.9...v0.15.0) (2024-05-12)

### Features

* Support go modules ([fae6506](https://github.com/prantlf/vp/commit/fae6506b8b04ab0b02e2d40973297d3bb1618d35))

## [0.14.9](https://github.com/prantlf/vp/compare/v0.14.8...v0.14.9) (2024-04-28)

### Bug Fixes

* Add missing linux-arm64 package ([b569867](https://github.com/prantlf/vp/commit/b56986709f6d475b847777b2e1d33ec30687aae8))

## [0.14.8](https://github.com/prantlf/vp/compare/v0.14.7...v0.14.8) (2024-04-17)

### Bug Fixes

* Rebuild with newer dependencies ([8551afe](https://github.com/prantlf/vp/commit/8551afec67bdd9dbbcc53c2809d6dd7249046484))

## [0.14.7](https://github.com/prantlf/vp/compare/v0.14.6...v0.14.7) (2024-04-17)

### Bug Fixes

* Log parsing and stringifying package.json and lock ([b627a12](https://github.com/prantlf/vp/commit/b627a12e776a54a3dd4543e048e4e67ee316cb86))

## [0.14.6](https://github.com/prantlf/vp/compare/v0.14.5...v0.14.6) (2024-04-14)

### Bug Fixes

* Work around with --bump-major-0 instead of -0 ([a11c22a](https://github.com/prantlf/vp/commit/a11c22a5182e1374d30dd40e9dc6cf938a927c8a))

## [0.14.5](https://github.com/prantlf/vp/compare/v0.14.4...v0.14.5) (2024-04-14)

### Bug Fixes

* Pass bump-major-0 to newchanges ([2a70724](https://github.com/prantlf/vp/commit/2a707244ab013c209473fe5f587c97aba51942dd))

## [0.14.4](https://github.com/prantlf/vp/compare/v0.14.3...v0.14.4) (2024-04-02)

### Bug Fixes

* Insert --no-verify before -am in git commit ([8854d50](https://github.com/prantlf/vp/commit/8854d50b329c36fc377b772da38b88133159f959))

## [0.14.3](https://github.com/prantlf/vp/compare/v0.14.2...v0.14.3) (2024-04-02)

### Bug Fixes

* Reword the description of --pre-release ([de4f673](https://github.com/prantlf/vp/commit/de4f67383559f90ee471d371b17e62a4407850aa))

## [0.14.2](https://github.com/prantlf/vp/compare/v0.14.1...v0.14.2) (2024-04-02)

### Bug Fixes

* Correct recognition of GitHub and GitLab URLs ([17e3f0a](https://github.com/prantlf/vp/commit/17e3f0a178b547f1c45b02a26a623728fda25cb2))

## [0.14.1](https://github.com/prantlf/vp/compare/v0.14.0...v0.14.1) (2024-03-31)

### Bug Fixes

* Enable github releases only for github repositories ([733cd84](https://github.com/prantlf/vp/commit/733cd84c6f8c921a36fab5e92d74b6cdbb1cec0f))

## [0.14.0](https://github.com/prantlf/vp/compare/v0.13.0...v0.14.0) (2024-03-30)

### Features

* Allow using --no-verify for disabling git hooks ([6ad6766](https://github.com/prantlf/vp/commit/6ad676674a9147cad1861f7b9dfdc449bf5da53b))
* Allow using -o ci.skip for disabling gitlab pipeline ([4a23cec](https://github.com/prantlf/vp/commit/4a23cec51934eddfb80e9f7b46e6ee2c6881c806))

### Bug Fixes

* Add "skip ci" to commit and tag mesages only for github ([99ac8ac](https://github.com/prantlf/vp/commit/99ac8acce83330d3e4483d7c1e358ee89adf43e5))

## [0.13.0](https://github.com/prantlf/vp/compare/v0.12.1...v0.13.0) (2024-03-30)

### Features

* Allow specifying a tag prefix ([818fc38](https://github.com/prantlf/vp/commit/818fc38d3a42b3d2aee2182d9a6d1a0ec1fffded))
* Support pre-releases ([e48d55b](https://github.com/prantlf/vp/commit/e48d55b663b16670c905fbb8bab97dbb855f731b))

## [0.12.1](https://github.com/prantlf/vp/compare/v0.12.0...v0.12.1) (2024-02-12)

### Bug Fixes

* Upgrade dependencies ([0d1f5e6](https://github.com/prantlf/vp/commit/0d1f5e603910c6e005a59020fcd834ba6fc27ba7))

## [0.12.0](https://github.com/prantlf/vp/compare/v0.11.1...v0.12.0) (2024-02-05)

### Features

* Allow removing "skip ci" from the versioning commits ([514a15d](https://github.com/prantlf/vp/commit/514a15dd156b2076bccd126b33539d70ca5ea878))
* Allow versioning a Rust cargo ([8a08026](https://github.com/prantlf/vp/commit/8a08026c1f0356840689d6c5f31dc5c77007f4ce))

## [0.11.1](https://github.com/prantlf/vp/compare/v0.11.0...v0.11.1) (2024-01-01)

### Bug Fixes


src/vp.v
import prantlf.cli { Cli, Env, run }

const version = '0.18.0'

const usage = 'Helps with development, installation and maintenance of VPM, NPM and Cargo packages.

Usage: vp [options] <command> [parameters]

Commands:
  init          generate a config file with defaults
  ls|list       list modules installed in ~/.vmodules or ./modules
  link          link the current module to the global ~./vmodules directory
  unlink        remove the current module link from the global ~./vmodules
  version       prepare the current module for publishing a new version
                (update changelog, bump version, commit and tag the change)
  publish       publish a new version prepared earlier by `vp version`
                (push the change and tag, create a github release)
  release       perform both `vp version` and `vp publish`

Options and parameters for list:
  -g|--global   inspect contents of ~./vmodules (default is ./[src/]modules)
  [<pkg names>] names of the package to print its version

Parameters for link and unlink:
  [<pkg name>]  name of the package if guessing is not reliable

Options for link and unlink:
  -f|--force    proceed even if the guessed package name was not reliable

Parameters for version and publish:
  [<version>]   version if the changelog update is disabled (also major,
                minor, patch or pre to bump the current version)

Options for version, publish and release:
  --no-changes        do not update the changelog
  --no-bump           do not bump the version in the package manifest
  --no-vlang          do not version and publish using v and v.mod
  --no-node           do not version and publish using npm and package.json
  --no-rust           do not version and publish using rust and Cargo.toml
  --no-golang         do not version and publish using go and go.mod
  --no-commit         do not commit the changes during publishing
  --no-commit-skip-ci do not add [skip ci] to the commit with the changes
  --no-tag            do not tag the commit during publishing
  --no-tag-skip-ci    do not add [skip ci] to the commit with the tag
  --no-push           do not push the commit and tag during publishing
  --no-push-skip-ci   do not add -o ci.skip to the push with the changes
  --no-release        do not create a new github release
  --no-archives       do not upload platform archives automatically as assets
  --no-failure        do not fail in case of no version change or release
  --no-verify         append --no-verify to commit and push
  --no-npm-auth       do not require npm token and write it to .npmrc
  --nc-args <args>    extra arguments for newchanges, enclosed in quotes
  --pre-release       publish a pre-release, not a stable release
  --pre-id <id>       first pre-release identifier (default: "next")
  -0|--bump-major-0   bump the major version also if it is 0
  -t|--tag-prefix <prefix> expect git tags prefixed (default: "v")
  -a|--assets <file>  files to upload as extra assets to the github release
  -b|--bump-files <file> extra files in which to bump the version
  -y|--yes            answer the push and release confirmations with "yes"
  -d|--dry-run        only print what would be done without doing it
  -v|--verbose        print the new changes on the console too

Common options:
  -c|--config <name>  file name of path of the config file
  -V|--version        print the version of the executable and exits
  -h|--help           print the usage information and exits

Examples:
  $ vp link prantlf.cargs -f
  $ vp unlink
  $ vp publish -v'

struct Opts {
	global          bool
	force           bool
	changes         bool = true
	bump            bool = true
	vlang           bool = true
	node            bool = true
	rust            bool = true
	golang          bool = true
	commit          ?bool
	commit_skip_ci  bool = true
	tag             ?bool
	tag_skip_ci     bool = true
	push            bool = true
	push_skip_ci    bool = true
	release         bool = true
	archives        bool = true
	failure         bool = true
	verify          bool = true
	npm_auth        bool = true
	nc_args         string @[json: 'nc-args']
	bump_major_0    bool   @[json: 'bump-major-0']
	pre_release     bool   @[json: 'pre-release']
	pre_id          string = 'next' @[json: 'pre-id']
	tag_prefix      string = 'v' @[json: 'tag-prefix']
	assets          []string
	bump_files      []string @[json: 'bump-files'; split]
	yes             bool
	dry_run         bool @[json: 'dry-run']
	verbose         bool
	version_detect  string = r'version' @[json: 'version-detect']
	version_replace string = r'(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(-[.0-9A-Za-z-]+)?' @[json: 'version-replace']
	gh_token        string @[json: 'gh-token']
	npm_token       string @[json: 'npm-token']
}

fn main() {
	run(Cli{
		usage:            usage
		version:          version
		options_anywhere: true
		cfg_opt:          'c'
		cfg_gen_arg:      'init'
		cfg_file:         '.vp'
		env:              Env.both
	}, body)
}

fn body(mut opts Opts, args []string) ! {
	if args.len == 0 {
		return error('Command is missing.')
	}

	command := args[0]
	first_arg := if args.len > 1 {
		args[1]
	} else {
		''
	}
	rest_args := if args.len > 1 {
		args[1..]
	} else {
		[]string{}
	}

	match command {
		'ls', 'list' {
			if opts.global {
				list_global(rest_args)!
			} else {
				list(rest_args)!
			}
		}
		'link' {
			link_path, module_dir := get_link(first_arg, opts.force)!
			link(link_path, module_dir, opts.force)!
		}
		'unlink' {
			link_path, module_dir := get_link(first_arg, opts.force)!
			unlink(link_path, module_dir, opts.force)!
		}
		'version' {
			commit := opts.commit or { true }
			tag := opts.tag or { true }
			create_version(first_arg, commit, tag, &opts)!
		}
		'publish' {
			commit := opts.commit or { false }
			tag := opts.tag or { false }
			publish(commit, tag, &opts)!
		}
		'release' {
			commit := opts.commit or { true }
			tag := opts.tag or { true }
			version_and_publish(first_arg, commit, tag, &opts)!
		}
		else {
			return error('Command "${command}" is invalid.')
		}
	}
}
